File: consts.py
from typing import Union

ObjType = Union[int, float, str, bool, bytes, list, tuple, dict, None]
ByteLike = Union[bytes, bytearray]
NUMBER_BASE = 255
ENDING_FLAG = 255
HEADER = b"<o>"
BIG_ENDIAN_DOUBLE_FORMAT = "!d"
BYTES_IN_DOUBLE = 8
NUMBER_OF_BITS_IN_BYTE = 8
LEFTMOST_BIT_MASK = 128


nums = iter(range(255))
POSITIVE_INT_FLAG = bytes([next(nums)])[0]
NEGATIVE_INT_FLAG = bytes([next(nums)])[0]
INT_FLAG = bytes([next(nums)])[0]
FLOAT_FLAG = bytes([next(nums)])[0]
STR_FLAG = bytes([next(nums)])[0]
EMPTY_STR_FLAG = bytes([next(nums)])[0]
BYTES_FLAG = bytes([next(nums)])[0]
EMPTY_BYTES_FLAG = bytes([next(nums)])[0]
BOOL_FLAG = bytes([next(nums)])[0]
TRUE_FLAG = bytes([next(nums)])[0]
FALSE_FLAG = bytes([next(nums)])[0]
NULL_FLAG = bytes([next(nums)])[0]
LIST_FLAG = bytes([next(nums)])[0]
EMPTY_LIST_FLAG = bytes([next(nums)])[0]
CONSISTENT_TYPE_LIST_FLAG = bytes([next(nums)])[0]
DICT_FLAG = bytes([next(nums)])[0]
EMPTY_DICT_FLAG = bytes([next(nums)])[0]
STR_KEY_DICT_FLAG = bytes([next(nums)])[0]
CONSISTENT_TYPE_DICT_FLAG = bytes([next(nums)])[0]
POINTER_FLAG = bytes([next(nums)])[0]

SMALL_INTS = {i: bytes([x])[0] for i, x in enumerate(nums)}
REVERSE_SMALL_INTS = {v: k for k, v in SMALL_INTS.items()}

File: src/utils/consts.rs
pub const NUMBER_BASE: u128 = 255;
pub const NUMBER_BASE_USIZE: usize = NUMBER_BASE as usize;
pub const ENDING_FLAG: u8 = 255;
pub const NEGATIVE_NUMBER_SIGN: u128 = NUMBER_BASE - 1;
pub const LEFTMOST_BIT_MASK: u8 = 128;
pub const HEADER: &[u8] = b"<o>";
pub const INVALID_UTF_8_START_BYTE: u8 = 0xff;

pub const EMPTY_STR_FLAG: u8 = 0;
pub const EMPTY_BYTES_FLAG: u8 = 1;
pub const TRUE_FLAG: u8 = 2;
pub const FALSE_FLAG: u8 = 3;
pub const NULL_FLAG: u8 = 4;
pub const EMPTY_LIST_FLAG: u8 = 5;
pub const EMPTY_DICT_FLAG: u8 = 6;
pub const POSITIVE_INT_FLAG: u8 = 7;
pub const NEGATIVE_INT_FLAG: u8 = 8;
pub const INT_FLAG: u8 = 9;
pub const FLOAT_FLAG: u8 = 10;
pub const STR_FLAG: u8 = 11;
pub const BYTES_FLAG: u8 = 12;
pub const BOOL_FLAG: u8 = 13;
pub const LIST_FLAG: u8 = 14;
pub const CONSISTENT_TYPE_LIST_FLAG: u8 = 15;
pub const DICT_FLAG: u8 = 16;
pub const STR_KEY_DICT_FLAG: u8 = 17;
pub const POINTER_FLAG: u8 = 18;

pub const AMOUNT_OF_USED_FLAGS: u8 = 30; // for future flags

pub const NOT_A_STR_BUT_A_POINTER_FLAG: [u8; 2] = [
    AMOUNT_OF_USED_FLAGS + 1,   // a str of length 1
    INVALID_UTF_8_START_BYTE,   // sike, not really
];
